unsupervised_rock_detection_3d/README.md

```diff
@@ -4,10 +4,52 @@
 
 This module extends the 3D Rock Detection project by focusing on unsupervised segmentation techniques for 3D rock point clouds. It emphasizes k-means clustering and region growing segmentation to enhance rock detection accuracy. The k-means clustering technique groups similar points based on their spatial features, while region growing relies on seed points and neighboring criteria to expand regions. These techniques enable precise segmentation of rock points from supporting surfaces, thus facilitating in-depth geometric analysis of geological formations.
 
+**Original and Segmented point cloud:**
+
+<img src="./images/Original.png" height="200"> <img src="./images/RegionGrowing2.png" height="200">
+
 ## Workflow
 
 The module involves two primary unsupervised segmentation techniques:
 
+### Region Growing Segmentation
+
+1. **Initialization**:
+
+    * Configure the RegionGrowingSegmentation class with the point cloud and segmentation parameters such as voxel size, number of neighbors, smoothness threshold, distance threshold, and curvature threshold.
+    * Downsample the point cloud using a voxel grid to reduce computation.
+    * Estimate normals for the downsampled point cloud to use in segmentation criteria. Normals are oriented consistently for reliable dot product calculations.
+    * Build a KD-Tree for efficient neighbor searches.
+
+2. **Seed Selection**:
+
+    * Compute the bounding box of the point cloud to find the geometric bounds.
+    * Determine the centroid of the bounding box in the x and y dimensions.
+    * Identify the point with the highest z value near the centroid, representing a potential rock seed.
+    * Identify the lowest point in the bounding box, representing a potential terrain seed.
+
+3. **Region Growing**:
+
+    * Precompute neighbors for each point within the specified distance threshold to speed up segmentation.
+    * Initialize regions from the selected seed points (highest point for rocks, bottommost point for terrain).
+    * Expand regions from seeds by evaluating neighbors based on the segmentation criteria (normal vector smoothness and curvature):
+        * Start with the seed point in a queue.
+        * For each point, evaluate its neighbors based on the segmentation criteria.
+        * Check if neighbors meet the smoothness and curvature thresholds. If they do, add them to the current region and the queue.
```

Comment:
you should add the mathematical definition of smoothness and curvature. Also, please clarify that we should just use one of them.

Reply:
Got it! Will add it and create another PR.